### PR TITLE
Enable `image-segmentation` on `AutoModelForSemanticSegmentation`

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Enable `image-segmentation` on `AutoModelForSemanticSegmentation`

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

@NielsRogge 
Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/pipelines/image_segmentation.py [hunk 1]
```diff
@@ -105,23 +111,46 @@ def _forward(self, model_inputs):
...
+                    f"Logits don't have expected dimensions, expected [1, {num_labels}, H, W], got {logits.shape}"
+                )
+            size = model_outputs["target_size"].tolist()[0]
+            logits_reshaped = torchvision.transforms.Resize(size)(logits)
+            classes = logits_reshaped.argmax(dim=1)[0]
+            annotation = []
+            for label_id in range(num_labels):
+                label = self.model.config.id2label[label_id]
+                mask = classes == label_id
+                score = ((mask * logits_reshaped).sum() / mask.sum()).exp().item()
```
Comment (Narsil) [inline src/transformers/pipelines/image_segmentation.py]: This score is the best I could come up on the spot.

In my estimation this is an average probability overall all pixels affected to that class. (using arithmetic mean of logits).
Is that sound ?

File: src/transformers/pipelines/image_segmentation.py [hunk 2]
```diff
@@ -105,23 +111,46 @@ def _forward(self, model_inputs):
...
+                    f"Logits don't have expected dimensions, expected [1, {num_labels}, H, W], got {logits.shape}"
+                )
+            size = model_outputs["target_size"].tolist()[0]
+            logits_reshaped = torchvision.transforms.Resize(size)(logits)
+            classes = logits_reshaped.argmax(dim=1)[0]
+            annotation = []
+            for label_id in range(num_labels):
+                label = self.model.config.id2label[label_id]
+                mask = classes == label_id
+                score = ((mask * logits_reshaped[0, label_id]).sum() / mask.sum()).exp().item()
```
Comment (Narsil) [inline src/transformers/pipelines/image_segmentation.py:151]: This score is the best I could come up on the spot.

In my estimation this is an average probability overall all pixels affected to that class. (using arithmetic mean of logits).
Is that sound ?

File: src/transformers/pipelines/image_segmentation.py [hunk 3]
```diff
@@ -105,23 +112,46 @@ def _forward(self, model_inputs):
...
+                for vals in zip(raw_annotation["scores"], raw_annotation["labels"], raw_annotation["masks"])
+            ]
+        else:
+            # Default logits
+            logits = model_outputs.logits
+            if len(logits.shape) != 4:
+                raise ValueError(f"Logits don't have expected dimensions, expected [1, N, H, W], got {logits.shape}")
+            # Softmax
+            logits = logits.log_softmax(dim=1)
+            b, n, h, w = logits.shape
```
Comment (sgugger) [inline src/transformers/pipelines/image_segmentation.py:139]: Better variable names please :-)

File: src/transformers/pipelines/image_segmentation.py [hunk 4]
```diff
@@ -105,23 +111,46 @@ def _forward(self, model_inputs):
...
+            num_labels = len(self.model.config.id2label)
+            if n != num_labels:
+                raise ValueError(
+                    f"Logits don't have expected dimensions, expected [1, {num_labels}, H, W], got {logits.shape}"
+                )
+            size = model_outputs["target_size"].tolist()[0]
+            logits_reshaped = torchvision.transforms.Resize(size)(logits)
```
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/pipelines/image_segmentation.py:146]: Is this for semantic segmentation? Logits sould have the same shape as the inputs/maks, so no resize should be needed.

File: src/transformers/pipelines/image_segmentation.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/pipelines/image_segmentation.py:152]: The score per pixel should be the softmax of the logits, I believe.

**Review by sgugger — COMMENTED** [review]
I think it's a mistake to group together tasks as different as instance segmentation and semantic segmentation in the same pipeline, especially when we haven't fully defined the instance segmentation outputs yet.

It's as if we grouped token classification and text classification in the same task for NLP, just because they both have classification in their names, in my opinion

It would be cleaner to have a separate `semantic-image-segmentation`.

File: src/transformers/pipelines/image_segmentation.py (hunk 2 above)
Comment (Narsil) [reply src/transformers/pipelines/image_segmentation.py:152]: Yes I use the log_softmax to get the logits.

File: src/transformers/pipelines/image_segmentation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/image_segmentation.py:146]: Well for the pipeline, the output expects the image to be in the original input size (the `feature_extractor` does a downsampling, so the pipeline needs to upsample again, because I don't expect pipeline users to do that on their own).

**Comment by Narsil** [comment]
> I think it's a mistake to group together tasks as different as instance segmentation and semantic segmentation in the same pipeline, especially when we haven't fully defined the instance segmentation outputs yet.

It's a long standing debate between us, and I think we just need to define what are pipelines really and have a proper design doc on which we agree on.
I think I understand your position, but I think we differ on the premises.


IMHO, what defines a pipeline is IN ORDER:

- Achieve a given task, targeted towards non-ml python programmers, hiding as many ML details without misleading users (transformers have a limited range of attention, so this shouldn't be hidden).
- A task, is a set of input/output. For simplicity, aim to core python types first `str`,  `dict`, `list`, `int`, `float`. Use "accepted standards" everywhere possible (`PIL.Image` for Python images for instance). 
- Compromise as little as possible performance wise.
- Parameters should enable advanced usage/tweaks only. Users shouldn't have to use them on standard usage.

That means:

- `AutoModelForXXX` 1-1 mapping with a task is not something which makes sense in general (`automatic-speech-recognition` does ASR both for CTC, and Seq2seq, and users shouldn't have to care about the difference between those two things).Input: `audio` -> output: `str` *is* what defines the pipeline here.




> 
> It's as if we grouped token classification and text classification in the same task for NLP, just because they both have classification in their names, in my opinion

Well `token-classification` is the same thing as `text-segmentation` actually (it's just not being used as a name by the community). If we follow the guidelines I suggest, the pipeline should actually never mention `tokens` (it's an ML concept). What non ML users should care about is "part of text" (because "words" is not necessarily what you are looking for

My personnal picture is that more generally, for an input set K ⊂ Ω:

classification are applications such that  `K -> [0; n]`
segmentation :  ` K -> {M,  M ⊂ K}` (parts of K)
generation: `K -> {K_i, K_i ⊂ Ω}` (new members of Ω)

This applies to Audio, Image, and Text pretty fine I think.

> 
> It would be cleaner to have a separate `semantic-image-segmentation`.

**Comment by sgugger** [comment]
> A task, is a set of input/output. For simplicity, aim to core python types first str, dict, list, int, float. Use "accepted standards" everywhere possible (PIL.Image for Python images for instance).

This is the crux of the problem. Instance segmentation and semantic segmentations have very different kinds of outputs. Grouping them in the same pipeline will be confusing to the user and it make the code unreadable, so I really don't see why we should group them in the same alias.

File: src/transformers/pipelines/image_segmentation.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sgugger) [reply src/transformers/pipelines/image_segmentation.py:158]: Ah forgot about that part. You'll need the resize indeed.

File: src/transformers/pipelines/image_segmentation.py (hunk 2 above)
Comment (sgugger) [reply src/transformers/pipelines/image_segmentation.py:164]: Sorry I was reading this wrong. I don't think we should return a score per label like this. It doesn't correspond to any metric I have seen on semantic segmentation, but I may have missed something. If you want to continue the analgoy with token classification being semantic segmentation, there should be one score per pixel, which is the probability corresponding to the predicted class.

File: src/transformers/pipelines/image_segmentation.py [hunk 5]
```diff
@@ -104,39 +119,67 @@ def _forward(self, model_inputs):
...
+            classes = logits_reshaped.argmax(dim=1)[0]
+            annotation = []
+            for label_id in range(num_labels):
+                label = self.model.config.id2label[label_id]
+                mask = classes == label_id
+                score = ((mask * logits_reshaped[0, label_id]).sum() / mask.sum()).exp().item()
+                mask = Image.fromarray(mask.numpy().astype(np.int8), mode="L")
+                if not raw_image:
+                    mask = self._get_mask_str(mask)
+                annotation.append({"score": score, "label": label, "mask": mask})
```
Comment (sgugger) [inline src/transformers/pipelines/image_segmentation.py:168]: This is where grouping the tasks under the same umbrella is dangerous. The output of the semantic segmentation should be one mask, not one per label.

**Comment by Narsil** [comment]
Major modifications with breaking changes linked to a good compromise we found with @sgugger .

My major opposition for creating a separate pipeline were:
- It adds complexity on non ML (users need to know about the differences between 200 tasks) with sometimes only minor differences in what a user actually cares for (where is the tumor located on the image, and how big is it ?)
- It forces users to use odd matrix `[[0, 2, 9], [0, 2, 2], ...]`  which are not real images but classes ids 2D array, which are confusing when you try to display them as they are pure black when displayed in colab for instance

@sgugger main opposition to a single pipeline
- We're trying to force feed different outputs in the same object where they are not really the same thing
  - Discussing further, the real problem is the confusion that stems from `parameters` which could be different for `Semantic` vs `Image segmentation` or `panoptic`. This is definitely an issue with parameters in general for pipelines.
  - `score` is a good example, in semantic segmentation it's not really meaningful, people mostly look at individual pixel logits, not really at the full cluster
- semantic segmentation can output a single mask for 2 persons, it could be confusing.

- Consensus, was to really aim on outputting at "non ML" objects.
Most ML users that want the details, should be able to use the raw AutoModelForXX, or even override the pipelines to modify subtle things. It might be something to explore, document more  in the future, but the core target for pipeline are non ML users, so we're ok giving up some something to get simplicity.
- `scores` don't make sense for semantic, so don't attempt to give them for semantic.

We're also working on a pipeline guidelines, to help guide us in decision, we'll work on this a bit internally, and release later on a wider audience to get community feedback.

```
- Removed the string compressed PNG string, that's a job for and API. `transformers` 's users
 stay in python land.
- Removed the `score` for semantic segmentation. It has hardly a meaning
  on its own in this context.
- Don't include the grayscale with logits for now (which could enable
  users to get a sense of confidence). Might be done later.
- Don't include the surface of the mask (could be used for sorting by
  users, to filter out small masks). It's already calculable, and
  it's easier to add later, than to add now and break later if we need.
```

File: tests/test_pipelines_image_segmentation.py [hunk 6]
```diff
@@ -62,34 +77,59 @@ def get_test_pipeline(self, model, tokenizer, feature_extractor):
...
+        self.assertEqual([{"score": ANY(float, type(None)), "label": ANY(str), "mask": ANY(Image.Image)}] * m, outputs)
+
+        if isinstance(image_segmenter.model, DetrForSegmentation):
+            # We need to test batch_size with images with the same size.
+            # Detr doesn't normalize the size of the images, meaning we can have
+            # 800x800 or 800x1200, meaning we cannot batch simply.
+            # We simply bail on this
+            batch_size = 1
+        else:
+            batch_size = 2
```
Comment (sgugger) [inline tests/test_pipelines_image_segmentation.py:109]: Nit, I would unindent the comment and put the whole test in one line.
```py
batch_size = 1 if isinstance(image_segmenter.model, DetrForSegmentation) else 2
```

**Review by sgugger — APPROVED** [review]
Thanks for updating the code!

File: src/transformers/pipelines/__init__.py [hunk 7]
```diff
@@ -256,7 +257,7 @@
     "image-segmentation": {
         "impl": ImageSegmentationPipeline,
         "tf": (),
-        "pt": (AutoModelForImageSegmentation,) if is_torch_available() else (),
+        "pt": (AutoModelForImageSegmentation, AutoModelForSemanticSegmentation) if is_torch_available() else (),
```
Comment (NielsRogge) [inline src/transformers/pipelines/__init__.py:260]:
```suggestion
        "pt": (AutoModelForImageSegmentation, AutoModelForSemanticSegmentation) if is_torch_available() else (),
```
The `AutoModelForImageSegmentation` class is likely to be renamed to `AutoModelForPanopticSegmentation`, cc @FrancescoSaverioZuppichini .

File: src/transformers/pipelines/__init__.py [hunk 8]
```diff
@@ -256,7 +257,7 @@
...
         "default": {"model": {"pt": "facebook/detr-resnet-50-panoptic"}},
```
Comment (NielsRogge) [inline src/transformers/pipelines/__init__.py:261]: If we only have a single pipeline for all tasks, we can also only have one default model :/

i.e. it's not possible to showcase semantic segmentation with a default one

File: src/transformers/pipelines/image_segmentation.py [hunk 9]
```diff
@@ -16,8 +14,13 @@
 
 if is_torch_available():
     import torch
+    from torch.nn import functional as F
```
Comment (NielsRogge) [inline src/transformers/pipelines/image_segmentation.py:17]:
```suggestion
    from torch import nn
```
In other parts of the Transformers library, we prefer to not use single-letter abbreviations. Rather, we use nn.functional in the code.

File: src/transformers/pipelines/image_segmentation.py [hunk 10]
```diff
@@ -77,16 +82,16 @@ def __call__(self, *args, **kwargs) -> Union[Predictions, List[Prediction]]:
...
             The dictionaries contain the following keys:
 
             - **label** (`str`) -- The class label identified by the model.
-            - **score** (`float`) -- The score attributed by the model for that label.
-            - **mask** (`str`) -- base64 string of a grayscale (single-channel) PNG image that contain masks
-              information. The PNG image has size (heigth, width) of the original image. Pixel values in the image are
-              either 0 or 255 (i.e. mask is absent VS mask is present).
+            - **mask** (`PIL.Image`) -- Pil Image with size (heigth, width) of the original image. Pixel values in the
+              image are in the range 0-255. 0 means the pixel is *not* part of the *label*, 255 means it definitely is.
+            - **score** (*optional* `float`) -- Optionnally, when the model is capable of estimating a confidence of
```
Comment (NielsRogge) [inline src/transformers/pipelines/image_segmentation.py:93]:
```suggestion
            - **score** (*optional* `float`) -- Optionally, when the model is capable of estimating a confidence of
```

File: src/transformers/pipelines/image_segmentation.py [hunk 11]
```diff
@@ -104,40 +109,56 @@ def _forward(self, model_inputs):
...
+                dict(zip(keys, vals))
+                for vals in zip(raw_annotation["scores"], raw_annotation["labels"], raw_annotation["masks"])
+            ]
+        else:
+            # Default logits
+            logits = model_outputs.logits
+            logits = logits.softmax(dim=1)
+            if len(logits.shape) != 4:
+                raise ValueError(f"Logits don't have expected dimensions, expected [1, N, H, W], got {logits.shape}")
+            # logits = logits.log_softmax(dim=1)
```
Comment (NielsRogge) [inline src/transformers/pipelines/image_segmentation.py:137]:
```suggestion
```
To be removed?

**Review by NielsRogge — APPROVED** [review]
Thanks for adding this. DETR is going to have a refactor after we add MaskFormer, so the `AutoModelForImageSegmentation` is going to be deprecated.

File: src/transformers/pipelines/__init__.py (hunk 8 above)
Comment (Narsil) [reply src/transformers/pipelines/__init__.py:269]: Defaults are always limited :(.

But I think users tend to mostly copy paste blogposts + doc examples so if we use `pipeline(model="nvidia/segformer-b5-finetuned-ade-640-640")` in those, it should help people discover them, wdyt ?

Comment (Narsil) [reply src/transformers/pipelines/__init__.py:269]: This is actually a good remark, that makes me think about pipelines vs tasks, and make we could imagine providing aliases `semantic-segmentation` that resolves to another default ? Food for thought.

**Comment by Narsil** [comment]
@sgugger Since this contains a breaking change, should I wait after the release the make the change or not ?

**Comment by sgugger** [comment]
I see this more as a bug fix than a breaking change, but let's see what @LysandreJik thinks.
Lysandre, for context, the pipeline used to return string for masks (so they can be used in the inference API) and it would return mask as PIL images now. I think it's completely okay since it makes the output of the pipeline more readable.

**Review by LysandreJik — COMMENTED** [review]
This is fine for me too as this was very recently introduced and is an experimental API.

Yay for the compromise and work on this, thanks @sgugger, @Narsil!
